tests/e2e/macros/mod.rs

```diff
@@ -0,0 +1,170 @@
+use std::{fmt::Display, ops::Not};
+
+use crate::support::{
+    MockClient,
+    cursor::peek_caret,
+    cursors,
+    fixture::Fixture,
+    normalize::{DiagnosticsWithUrl, normalize_diagnostics},
+    sandbox,
+};
+use cairo_lang_utils::ordered_hash_map::OrderedHashMap;
+use cairo_language_server::lsp::ext::ExpandMacro;
+use itertools::Itertools;
+#[cfg(doc)]
+use itertools::Itertools;
+use lsp_types::{Position, TextDocumentIdentifier, TextDocumentPositionParams};
+use serde::Serialize;
+use serde_json::json;
+
+mod fixtures;
+mod procedural;
+
+/// Test report. Contains all diagnostics sent by the language server in the test project
+/// and all macro expansions collected during the test.
+#[derive(Debug, Serialize)]
+pub struct Report {
+    expansions: Option<Vec<ExpansionGroup>>,
+    mapped_diagnostics: Option<Vec<MappedDiagnostic>>,
+}
+
+impl Report {
+    fn new(expansions: Vec<ExpansionGroup>, mapped_diagnostics: Vec<MappedDiagnostic>) -> Self {
+        let expansions = expansions.is_empty().not().then_some(expansions);
+        let mapped_diagnostics = mapped_diagnostics.is_empty().not().then_some(mapped_diagnostics);
+        Self { mapped_diagnostics, expansions }
+    }
+}
+
+impl Display for Report {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            toml::to_string_pretty(&self).expect("Report should be serializable to TOML.")
+        )
+    }
+}
+
+/// Macro expansion obtained by calling the LSP method
+/// on a set of lines in the source code,
+/// collectively represented as `analyzed_lines`.
+#[derive(Debug, Serialize)]
+pub struct ExpansionGroup {
+    analyzed_lines: String,
+    generated_code: String,
+}
+
+/// Representation of the [`lsp_types::Diagnostic`] mapped to the source code with additional URL,
+/// suitable for TOML serialization.
+#[derive(Debug, Serialize)]
+pub struct MappedDiagnostic {
+    url: String,
+    range: String,
+    message: String,
+    severity: Option<String>,
+}
+
+impl MappedDiagnostic {
+    pub fn new(diagnostic: lsp_types::Diagnostic, url: String) -> Self {
+        let severity = diagnostic.severity.map(|severity| format!("{severity:?}"));
+
+        let lsp_types::Range { start, end } = diagnostic.range;
+        let range = format!("{}:{}-{}:{}", start.line, start.character, end.line, end.character);
+
+        Self { url, range, message: diagnostic.message, severity }
+    }
+}
+
+/// An environment for testing macros.
+pub trait MacroTestFixture {
```

Comment:
```suggestion
pub trait MacroTest {
```

to follow the naming I used in `GotoDefinitionTest`

Reply:
Done